endtoend/helpers.go

```diff
@@ -0,0 +1,109 @@
+package endtoend
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+func killProcesses(t *testing.T, pIDs []int) {
+	for _, id := range pIDs {
+		process, err := os.FindProcess(id)
+		if err != nil {
+			t.Fatalf("Could not find process %d: %v", id, err)
+		}
+		if err := process.Kill(); err != nil {
+			t.Fatal(err)
+		}
+		if err := process.Release(); err != nil {
+			t.Fatal(err)
+		}
+	}
+}
+
+func waitForTextInFile(file *os.File, text string) error {
```

Comment:
We should always avoid passing in file strings into functions in Go. Use the helpful and awesome `io.Reader` or `io.Scanner` interfaces Go provides and pass that interface in as a func arg to `waitForText` to do this in a more idiomatic way and test it! Here's a legendary resource I use for all file related items in Go: https://www.devdungeon.com/content/working-files-go and https://dev.to/dayvonjersen/how-to-use-io-reader-and-io-writer-ao0

Reply:
This also allows you to write tests for this function, otherwise it would be nearly impossible to test without mocking the file system itself

Reply:
Got it, thank you!